doc/man/man1/policygen.1
.TH policygen "1" "PPACK Commands" "November 2020" "PPACK Commands"

.SH "NAME"
policygen \-

.SH "SYNOPSIS"
\fBpolicygen\fR [\fIOPTIONS\fR...]

.SH "DESCRIPTION"
.PP
Generate masks that meet certain requirements to attack passwords, instead
of use brute force to attack them.

.SH "OPTIONS"

.TP "10"
\f3\-o\fP\f3,\fP \f3\-\-output\fP
This file contains all the generated masks that meet all requirements that we have entered.
.IP

.TP
\f3\-i \fP\f3, \fP \f3\-\-input\fP\
Input File. This one contains all the requiements for policygen, so this file can be used instead of all the requirement flags.
.IP


.TP
\f3\-q\fP\f3,\fP \f3\-\-quiet\fP
Ocult the PPACK's banner.
.IP

.TP
\f3\-\-pretty\fP
Print the result in a pretty format.
.IP

.TP
\fP \f3\-\-show\fP
Show the generated masks (bases).
.IP

.TP
\fP \f3\-\-minlength\fP
Minimum masks' length.
.IP

.TP
\fP \f3\-\-maxlength\fP
Maximum masks' length.
.IP


.TP
\fP \f3\-\-minlower\fP
Minimum number of lowercase characters.
.IP

.TP
\fP \f3\-\-maxlower\fP
Maximum number of lowercase characters.
.IP


.TP
\fP \f3\-\-minupper\fP
Minimum number of uppercase characters.
.IP

.TP
\fP \f3\-\-maxupper\fP
Maximum number of uppercase characters.
.IP


.TP
\fP \f3\-\-mindigit\fP
Minimum number of digits.
.IP

.TP
\fP \f3\-\-maxdigits\fP
Maximum number of digits.
.IP


.TP
\fP \f3\-\-minspecial\fP
Minimum number of special characters.
.IP

.TP
\fP \f3\-\-maxspecial\fP
Maximum number of special characters.
.IP


.SH "EXAMPLES"
This examples ilustrate the use of \f3policygen\fP requirements':
.RS
.PP
.nf
.ft 3
$ policygen --minlength 3 --maxlength 4 --minlower 2 --maxspecial 3 -o custom_attack.hcmasks


.ft 1
.fi
.RE
.PP
This example ilustrate how to enter  \f3policygen\fP requiments using a \f3json\fP file, because
sometime there are a lot of requirements:
.RS
.PP
.nf
.ft 3
$ policygen -i filter.ini

.PP
\fIrequiements.json\fR
        [policygen]
        OUTPUT="hard.hcmask"
        QUIET=false
        SHOW=true
        MINLENGTH=7
        MAXLENGTH=9
        MINLOWER=1
        MINUPPER=1
        MINSPECIAL=2
        MAXSPECIAL=3

.TP
\f3Note:\fP
.PP
The ini file no necessary will hace an unique section \fIpolicygen\fR instead it can have diferent sections but \f3policygen\fPis the unique section that will be read.


.ft 1
.fi
.RE
.PP
This example ilustrate how to run policygen with 4 threads\fIopenmp\fR:
.RS
.PP
.nf
.ft 3
# export OMP_NUM_THREADS=4
$ policygen --minlength 3 --maxlength 4 --minlower 2 --maxspecial 3 -o custom_attack.hcmasks
$ policygen -i filter.ini


src/policygen.cpp
/*
 * policygen - cli interface
 *
 * Depuration - 28 nov 2020
 * policygen (cli interface) depurate.
 *
 * No tested functions: ALL FUNCTION TESTED
 *
 * No implemented functions:
 *  
 * NO COMPLETED
 * --- implement the input of parameters using a input file ---
 *
 * NO COMPLETED
 * --- Implement a beatiful output (call helper function  in corePolicygen) ---
 *
 * NO COMPLETED
 * --- Implement mask writing in a file(call helper function  in corePolicygen) ---
 *
 *
 *
 * Maintainer: glozanoa <glozanoa@uni.pe>
 *
 */

#ifndef __INCLUDE_STD_IOSTREAM_H__
#define __INCLUDE_STD_IOSTREAM_H__
#include <cstdlib>
#include <iostream>
#endif //__INCLUDE_STD_IOSTREAM_H__

#ifndef __INCLUDE_PROGRAM_OPTIONS_H__
#define __INCLUDE_PROGRAM_OPTIONS_H__
#include<boost/program_options.hpp>
#endif //__INCLUDE_PROGRAM_OPTIONS_H__

using namespace boost;
namespace po = boost::program_options;


#ifndef __INCLUDE_PPACK_H__
#define __INCLUDE_PPACK_H__
#include "../include/ppack.hpp"
#endif // __INCLUDE_PPACK_H__

using namespace std;

int main(int argc ,char* argv[])
{

    //cout << "Start running policygen main" << endl;

    try
    {
    po::options_description files("Files I/O");
    files.add_options()
        ("output,o", po::value<string>()->default_value(""), "Ouput File.")
        ("input,i", po::value<string>()->default_value(""), "Input File.");

    po::options_description print("Print");
    print.add_options()
        ("show,s", po::value<bool>()->implicit_value(true)->default_value(false), "Show generated masks.")
        ("quiet,q", po::value<bool>()->implicit_value(true)->default_value(false), "Quiet printing(Omit PPACK logo).")
        ("pretty,p", po::value<bool>()->implicit_value(true)->default_value(false), "Pretty output.");


    po::options_description mask("Mask Structure");

    /*
    * MEANS of -1 value as a default value:
    *
    * EXAMPLES:
    * (we use lower as a concrete case, but this can be used equal for other cases[upper, digit and special also length])
    *
    * (if minlower=3 and (maxlower isn't entered)maxlower=-1 this means that the generated mask can have from 3 to more lower alpha)
    *
    * (if minlower and maxlower isn't entered: then minlower=0 and maxlower=0  that means that the password haven't loweralpha charactes), we create this rule to avoid enter unnecesary flags as --maxlower=0 , so instead of that you do not have to enter the flag related to lowers
    *
    *
    *
    * we create this rule to avoid enter unnecesary flags as --maxlower=0 , so instead of that you do not have to enter the flag related to lowers
    */


    mask.add_options()
        ("minlength", po::value<unsigned int>()->default_value(0), "Minimum password length.")
        ("maxlength", po::value<int>()->default_value(0), "Maximum password length.")

        ("minlower", po::value<unsigned int>()->default_value(0), "Minimum password lowercase characters.")
        ("maxlower", po::value<int>()->default_value(0), "Maximum password lowercase characters.")

        ("minupper", po::value<unsigned int>()->default_value(0), "Minimum password uppercase characters.")
        ("maxupper", po::value<int>()->default_value(0), "Maximum password uppercase characters.")


        ("mindigit", po::value<unsigned int>()->default_value(0), "Minimum password digits characters.")
        ("maxdigit", po::value<int>()->default_value(0), "Maximum password digits characters.")

        ("minspecial", po::value<unsigned int>()->default_value(0), "Minimum password special characters.")
        ("maxspecial", po::value<int>()->default_value(0), "Maximum password special characters.");

    po::options_description parallel("Parallel");
    parallel.add_options()
        ("threads,t", po::value<unsigned int>()->default_value(2), "Number of OMP threads.");
      

    po::options_description policygen("Generate customized mask for crack passwords");
    policygen.add_options()
        ("version,v", "PPACK version.")
        ("help,h", "Show help.");
    policygen.add(files).add(print).add(mask).add(parallel);

    po::variables_map vm;

    po::store(po::command_line_parser(argc, argv).
            options(policygen).run(), vm);

    if(vm.count("help")){
        cout << policygen << endl;
        exit(EXIT_FAILURE);
    }

    if(vm.count("version"))
    {
        //cout << "PPACK  version " + ppack::VERSION << endl;
        cout << "PPACK  version: 1.0"  << endl;
        exit(EXIT_FAILURE);
    }

    //cout << "before run pstruct constructor." << endl;
    pstruct pargs(vm, policygen);
    //pargs.debug();
    PPACK::policygen(pargs);

    return 0;
    }
    catch(std::exception& e)
    {
        cout << e.what() << endl;
        exit(EXIT_FAILURE);
    }
}
